File: day5/day5.py
#!/usr/bin/env python3

def count_steps(txt, strange):
    lst = list(map(int, txt.splitlines()))
    steps = 0
    i = 0
    size = len(lst)
    while 0 <= i < size:
        n = lst[i]
        if strange and n >= 3:
            lst[i] -= 1
        else:
            lst[i] += 1

        i += n
        steps += 1

    return steps

# assert count_steps("""0
# 3
# 0
# 1
# -3""", False) == 5
# assert count_steps("""0
# 3
# 0
# 1
# -3""", True) == 10

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", count_steps(txt, False))
    print("Part2:", count_steps(txt, True))

File: day5/day5.cpp
#include <fstream>
#include <iostream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static size_t count_steps(vector<int> offsets, bool strange)
{
	size_t steps = 0;
	int i = 0;
	while (0 <= i && (size_t)i < offsets.size())
	{
		int n = offsets[i];
		if (strange && n >= 3)
		{
			offsets[i]--;
		}
		else
		{
			offsets[i]++;
		}

		i += n;
		steps++;
	}
	return steps;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	vector<int> offsets;
	int offset;
	while (input>>offset)
	{
		offsets.push_back(offset);
	}
	input.close();

	cout << fmt::format("Part1: {}", count_steps(offsets, false)) << endl
	     << fmt::format("Part2: {}", count_steps(offsets, true)) << endl;
	return 0;
}
